### PR TITLE
Abstract Supabase from database query calls

### PR DESCRIPTION
**Pull request by raoufchebri**
This PR abstracts Supabase database code from the Quivr database calls.
This removes Quivr's dependency on Supabase database and allows the app to integrate other databases.

### CODE REVIEW

File: backend/models/databases/database.py [hunk 1]
```diff
@@ -1,19 +1,48 @@
 from abc import ABC, abstractmethod
+from uuid import UUID
 
 
 class Database(ABC):
```
Comment (Noezor) [inline backend/models/databases/database.py]: We should use typing as much as possible in the ABC I think

File: backend/models/databases/supabase/supabase.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+                item["brain_id"] for item in associated_brains_response.data
+            ]
+
+            # If the vector is not associated with any other brains, delete it from 'vectors' table
+            if not associated_brains:
+                self.db.table("vectors").delete().filter(
+                    "id", "eq", vector_id
+                ).execute()
+
+        return {"message": f"File {file_name} in brain {brain_id} has been deleted."}
```
Comment (Noezor) [inline backend/models/databases/supabase/supabase.py]: How much do you want to change the app logic? Ideally the `Database` only does database level operations, and the logic of the message is left to other applicative layers

However, I think there are strong arguments to leave the core logic as much as possible unchanged and we can do this in a subsequent MR

**Comment by Noezor** [comment]
LGTM now!

**Comment by Noezor** [comment]
https://github.com/raoufchebri/quivr/pull/1#discussion_r1267039026 I think this comment is still valid w.r.t. typing. The repository still does not return consistent types through the use of supabase (is this just a dict?). Maybe we want to wrap them in actual objects

**Comment by raoufchebri** [comment]
Screenshot 2023-07-20 at 00 11 29

File: .github/workflows/backend-tests.yml [hunk 3]
```diff
@@ -2,13 +2,33 @@ name: Backend Tests
...
+        type: choice
+        options:
+        - info
+        - warning
+        - debug
+      tags:
+        description: 'Test scenario tags'
+        required: false
+        type: boolean
+      environment:
```
Comment (Noezor) [inline .github/workflows/backend-tests.yml:27]: can we have example of values here?

File: backend/auth/api_key_handler.py [hunk 4]
```diff
@@ -13,14 +13,8 @@ async def verify_api_key(
...
         commons = common_dependencies()
-        result = (
-            commons["supabase"]
-            .table("api_keys")
-            .select("api_key", "creation_time")
-            .filter("api_key", "eq", api_key)
-            .filter("is_active", "eq", True)
-            .execute()
-        )
+        result = commons["db"].get_active_api_key(api_key)
```
Comment (Noezor) [inline backend/auth/api_key_handler.py:16]: We follow the use of `commons` here, but I am not sure pushing the db object in a dict enables the VSCode python server to be helpful and know the type of the variable?

File: backend/models/databases/supabase/brains.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+                item["brain_id"] for item in associated_brains_response.data
+            ]
+
+            # If the vector is not associated with any other brains, delete it from 'vectors' table
+            if not associated_brains:
+                self.db.table("vectors").delete().filter(
+                    "id", "eq", vector_id
+                ).execute()
+
+        return {"message": f"File {file_name} in brain {brain_id} has been deleted."}
```
Comment (Noezor) [inline backend/models/databases/supabase/brains.py:179]: Let's change this logic in another MR to instead have the methods of the repo not returning objects instead returning true/false or errors? 
The message here feels like app logic and probably should not be in the repository. 

The message is also wrong when the method is called on a filename/braid_id that does not exist

File: scripts/pg_tables.sql [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+CREATE TABLE IF NOT EXISTS migrations (
+  name VARCHAR(255)  PRIMARY KEY,
+  executed_at TIMESTAMPTZ DEFAULT current_timestamp
+);
+
+INSERT INTO migrations (name) 
+SELECT '202307111517031_change_vectors_id_type'
+WHERE NOT EXISTS (
+    SELECT 1 FROM migrations WHERE name = '202307111517031_change_vectors_id_type'
+);
```
Comment (Noezor) [inline scripts/pg_tables.sql:179]: new line at end of file

**Review by Noezor — APPROVED** [review]
I think it would be great to return typed objects in the repository to make it easier to have all the repositories match in output. Here, supabase does not always return objects in dict that match existing typed objects, and due to lack of typing of the response type this is ambiguous.

(can be left to another MR, this one is already significant)

--
Approving because my comments are pretty minor

File: .github/workflows/backend-tests.yml (hunk 3 above)
Comment (raoufchebri) [reply .github/workflows/backend-tests.yml:27]: I added that for testing purposes. Will likely remove it before merge

File: backend/models/databases/supabase/brains.py (hunk 5 above)
Comment (raoufchebri) [reply backend/models/databases/supabase/brains.py:179]: I'll change this. I missed it.

**Comment by raoufchebri** [comment]
I didn't want to strongly type the return value in case it's different from other integrations: in our case sqlalchemy

File: backend/auth/api_key_handler.py (hunk 4 above)
Comment (raoufchebri) [reply backend/auth/api_key_handler.py:16]: I agree. I suspect this refactoring should be on a PR of its own

**Pull request merged**